File: java/src/main/java/com/example/vulnapp/vulns/SsrfController.java
package com.example.vulnapp.vulns;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

/**
 * Server-Side Request Forgery (SSRF) —
 * OWASP 2021 A10 Server-Side Request Forgery  ->  folded into
 * OWASP 2025 A01 Broken Access Control.
 *
 * Mirrors the comment/annotation format of {@link SqlInjectionController}: each
 * permutation carries a standard header describing the flaw, its CWE, an
 * exploitation-likelihood rating, greenfield-vs-legacy prevalence, an example
 * payload, and the corresponding safe pattern. Every dangerous sink is a real,
 * exploitable {@link HttpURLConnection} that the SERVER opens to a
 * user-controlled URL — nothing is stubbed. Timeouts are short (~3s) so the
 * demo never hangs, and {@code HttpURLConnection} follows 3xx redirects by
 * default, which matters for the P3 redirect bypass.
 *
 * PRIMARY CWE: CWE-918 Server-Side Request Forgery (SSRF).
 */
@RestController
@RequestMapping("/ssrf")
public class SsrfController {

    private static final int TIMEOUT_MS = 3000;

    // ========================================================================
    // PERMUTATION 1 — Unrestricted server-side fetch (no allow-list)
    // OWASP 2021 A10 Server-Side Request Forgery  ->  folded into OWASP 2025 A01 Broken Access Control
    // CWE-918: Server-Side Request Forgery (SSRF)
    // EXPLOITATION LIKELIHOOD: HIGH — reachable pre-auth, deterministic, no
    //   tooling required. The server opens a GET to whatever URL the attacker
    //   supplies and returns status + body, so it doubles as an internal proxy:
    //   read cloud metadata (169.254.169.254), hit admin-only internal services,
    //   or use file:// / gopher:// depending on the client stack.
    // PREVALENCE TODAY: greenfield MEDIUM (URL-fetch features — link previews,
    //   image proxies, PDF/HTML renderers, import-from-URL — are ubiquitous and
    //   NO mainstream framework blocks SSRF by default; egress controls + IMDSv2
    //   are what save modern shops, not the code) | legacy/10yr tech-debt HIGH
    //   (hand-rolled fetchers with zero egress validation are everywhere).
    // EXAMPLE:
    //   curl 'http://localhost:8080/ssrf/fetch?url=http://169.254.169.254/latest/meta-data/iam/security-credentials/'
    //   curl 'http://localhost:8080/ssrf/fetch?url=http://127.0.0.1:8080/ssrf'
    // FIX: allow-list scheme + host and block private/link-local ranges after
    //   DNS resolution (see /ssrf/fetch-safe).
    // ========================================================================
    @GetMapping("/fetch")
    public Map<String, Object> fetch(@RequestParam(defaultValue = "http://example.com") String url) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("permutation", "unrestricted-fetch");
        // VULNERABLE: user-supplied URL fetched by the server with no allow-list / egress check.
        out.putAll(httpGet(url));
        return out;
    }

    // ========================================================================
    // PERMUTATION 2 — Blind / internal SSRF via a user-supplied webhook URL
    // OWASP 2021 A10 Server-Side Request Forgery  ->  folded into OWASP 2025 A01 Broken Access Control
    // CWE-918: Server-Side Request Forgery (SSRF)
    // EXPLOITATION LIKELIHOOD: HIGH — "register a callback/webhook and we'll POST
    //   to it" is a first-class feature, so the request is attacker-controlled by
    //   design. Even when the body is not returned (truly blind), timing / status
    //   codes turn the server into an internal PORT SCANNER, and pointing it at
    //   http://169.254.169.254/latest/meta-data/ steals cloud instance credentials.
    // PREVALENCE TODAY: greenfield MEDIUM (webhooks/callbacks are standard SaaS
    //   plumbing; SSRF-aware teams validate the destination, but many ship it raw
    //   because "it's just an outbound call") | legacy/10yr tech-debt HIGH
    //   (integration/notification subsystems that fire arbitrary stored URLs).
    // EXAMPLE:
    //   curl -X POST http://localhost:8080/ssrf/webhook -H 'Content-Type: application/json' \
    //        -d '{"url":"http://169.254.169.254/latest/meta-data/"}'
    //   curl -X POST http://localhost:8080/ssrf/webhook -H 'Content-Type: application/json' \
    //        -d '{"url":"http://127.0.0.1:22"}'   # internal port probe via status/timing
    // FIX: resolve the host and reject private/link-local targets; pin an
    //   allow-list of destinations (see /ssrf/fetch-safe).
    // ========================================================================
    @PostMapping("/webhook")
    public Map<String, Object> webhook(@RequestBody(required = false) Map<String, Object> body) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("permutation", "blind-webhook");
        Object raw = body == null ? null : body.getOrDefault("url", body.get("callback"));
        String url = raw == null ? "" : raw.toString();
        // VULNERABLE: server fires a request at an attacker-chosen callback URL with no destination check.
        out.putAll(httpGet(url));
        return out;
    }

    // ========================================================================
    // PERMUTATION 3 — Naive substring blocklist (trivially bypassable)
    // OWASP 2021 A10 Server-Side Request Forgery  ->  folded into OWASP 2025 A01 Broken Access Control
    // CWE-918: Server-Side Request Forgery (SSRF)
    // EXPLOITATION LIKELIHOOD: HIGH — the only control is a case-sensitive
    //   substring check for "localhost" / "127.0.0.1", which does not understand
    //   URL parsing, alternate IP encodings, DNS, IPv6, or redirects. Every one of
    //   the following reaches loopback/internal despite the filter:
    //     http://127.0.0.1.nip.io/     (DNS name that resolves to 127.0.0.1)
    //     http://0177.0.0.1/           (octal 0177 = 127)
    //     http://2130706433/           (decimal form of 127.0.0.1)
    //     http://[::1]/                (IPv6 loopback — no "127.0.0.1" substring)
    //     http://169.254.169.254/...   (metadata IP was never on the blocklist)
    //     an external URL that 302-redirects to http://127.0.0.1/ (follow-redirect)
    // PREVALENCE TODAY: greenfield LOW (a reviewed modern shop reaches for a
    //   vetted SSRF library / egress proxy, not a string blocklist; SAST flags
    //   deny-list URL checks) | legacy/10yr tech-debt HIGH (hand-rolled
    //   "block localhost" string checks are a classic band-aid that ships and stays).
    // EXAMPLE:
    //   curl 'http://localhost:8080/ssrf/preview?url=http://127.0.0.1.nip.io:8080/ssrf'
    //   curl 'http://localhost:8080/ssrf/preview?url=http://2130706433:8080/ssrf'
    //   curl 'http://localhost:8080/ssrf/preview?url=http://169.254.169.254/latest/meta-data/'
    // FIX: do NOT deny-list strings — resolve the host to IP(s) and allow only
    //   public destinations on an allow-list of schemes/hosts (see /ssrf/fetch-safe).
    // ========================================================================
    @GetMapping("/preview")
    public Map<String, Object> preview(@RequestParam(defaultValue = "http://example.com") String url) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("permutation", "naive-blocklist");
        // Broken control: a raw substring blocklist that ignores octal/decimal/IPv6/DNS/redirects.
        if (url.contains("localhost") || url.contains("127.0.0.1")) {
            out.put("blocked", true);
            out.put("reason", "url contains 'localhost' or '127.0.0.1'");
            out.put("requested_url", url);
            return out;
        }
        // VULNERABLE: passes the blocklist yet still reaches loopback/internal via encoding/DNS/redirect.
        out.putAll(httpGet(url));
        return out;
    }

    // ========================================================================
    // SAFE REFERENCE — scheme + host allow-list AND private/link-local block.
    // Shown so the vulnerable/safe pair can be diffed by SAST tooling and learners.
    //
    // Three independent controls make this safe:
    //   1) the scheme must be http/https (no file:, gopher:, ftp:, dict:, ...);
    //   2) the host must be on an explicit allow-list of destinations;
    //   3) EVERY resolved IP for that host is checked and rejected if it falls in
    //      loopback 127.0.0.0/8, private 10/8 · 172.16/12 · 192.168/16, or
    //      link-local 169.254/16 — closing the DNS-rebinding / alternate-encoding
    //      holes that P3's string blocklist leaves open. Only after all three
    //      pass is the request actually issued.
    // ========================================================================
    private static final Set<String> ALLOWED_SCHEMES = Set.of("http", "https");
    private static final Set<String> ALLOWED_HOSTS = Set.of("example.com", "www.example.com", "api.github.com");

    @GetMapping("/fetch-safe")
    public Map<String, Object> fetchSafe(@RequestParam(defaultValue = "http://example.com") String url) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("permutation", "safe-allowlist");
        out.put("requested_url", url);
        try {
            URL parsed = new URL(url);
            String scheme = parsed.getProtocol() == null ? "" : parsed.getProtocol().toLowerCase();
            String host = parsed.getHost() == null ? "" : parsed.getHost().toLowerCase();
            if (!ALLOWED_SCHEMES.contains(scheme)) {
                out.put("rejected", "scheme not in allow-list " + ALLOWED_SCHEMES);
                return out;
            }
            if (!ALLOWED_HOSTS.contains(host)) {
                out.put("rejected", "host not in allow-list " + ALLOWED_HOSTS);
                return out;
            }
            for (InetAddress addr : InetAddress.getAllByName(host)) {
                if (addr.isLoopbackAddress() || addr.isAnyLocalAddress()
                        || addr.isSiteLocalAddress() || addr.isLinkLocalAddress()) {
                    // SAFE: refuse any host that resolves into a private/link-local range.
                    out.put("rejected", "host resolves to a private/link-local address: " + addr.getHostAddress());
                    return out;
                }
            }
            // SAFE: scheme, host, and every resolved IP all passed validation before the fetch.
            out.putAll(httpGet(url));
        } catch (Exception e) {
            out.put("error", e.getClass().getSimpleName() + ": " + e.getMessage());
        }
        return out;
    }

    /**
     * Opens a real server-side GET to {@code urlStr} with short connect/read
     * timeouts and returns the status line plus a bounded body snippet. On any
     * failure it returns the exact URL that was requested together with the error
     * (as the reference modules echo the built command/query), so blind/internal
     * probes still surface a useful signal. {@code HttpURLConnection} follows 3xx
     * redirects by default — deliberately left on for the P3 redirect bypass.
     */
    private Map<String, Object> httpGet(String urlStr) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("requested_url", urlStr);
        HttpURLConnection conn = null;
        try {
            URL url = new URL(urlStr);
            conn = (HttpURLConnection) url.openConnection();
            conn.setConnectTimeout(TIMEOUT_MS);
            conn.setReadTimeout(TIMEOUT_MS);
            conn.setRequestMethod("GET");
            int status = conn.getResponseCode();
            out.put("status", status);
            out.put("final_url", conn.getURL().toString());
            InputStream in = status >= 400 ? conn.getErrorStream() : conn.getInputStream();
            String bodySnippet = "";
            if (in != null) {
                byte[] raw = in.readAllBytes();
                in.close();
                String body = new String(raw, StandardCharsets.UTF_8);
                bodySnippet = body.length() > 500 ? body.substring(0, 500) : body;
            }
            out.put("body_snippet", bodySnippet);
        } catch (Exception e) {
            out.put("error", e.getClass().getSimpleName() + ": " + e.getMessage());
        } finally {
            if (conn != null) {
                conn.disconnect();
            }
        }
        return out;
    }
}

File: csharp/Controllers/SsrfController.cs
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Threading.Tasks;

namespace VulnApp.Controllers;

/// <summary>
/// Server-Side Request Forgery (SSRF) —
/// OWASP 2021 A10 Server-Side Request Forgery -> folded into OWASP 2025 A01 Broken Access Control.
///
/// Mirrors the annotation format and attribute-routed action style of
/// SqlInjectionController. Every handler drives a REAL HttpClient request against
/// a user-controlled URL, so the forgery is genuinely exploitable when the app
/// runs. The canonical high-impact target is the cloud metadata service at
/// http://169.254.169.254/latest/meta-data/ (IMDSv1 hands out IAM credentials with
/// no auth); a target that works even inside this sandbox is the app's OWN internal
/// admin endpoint, e.g. http://127.0.0.1:5001/access/admin/users — the server can
/// be coerced into requesting internal resources the client could never reach.
///
/// There is no "safe by default" framework for SSRF: fetching a URL is the intended
/// feature (webhooks, link previews, image proxies, PDF/HTML importers, "import from
/// URL"). Safety comes from validating the RESOLVED destination, not the string.
/// All fetches use a ~3s timeout so the demo never hangs.
/// </summary>
[ApiController]
public class SsrfController : ControllerBase
{
    // How long any outbound request may take before we give up (keeps the demo snappy).
    private static readonly TimeSpan FetchTimeout = TimeSpan.FromSeconds(3);

    // ========================================================================
    // PERMUTATION 1 — direct fetch of a fully user-controlled URL (no allow-list)
    // OWASP 2021 A10 Server-Side Request Forgery -> folded into OWASP 2025 A01 Broken Access Control
    // CWE-918: Server-Side Request Forgery (SSRF)
    // EXPLOITATION LIKELIHOOD: HIGH — unauthenticated GET, zero tooling; the server
    //   dials any scheme/host/IP the caller names. Trivially reaches internal-only
    //   services and cloud metadata; ultimate impact is env-dependent (IMDSv1 vs v2,
    //   what internal endpoints exist) but reaching the sink is deterministic.
    // PREVALENCE TODAY: greenfield MEDIUM (nothing makes "fetch a URL" safe by
    //   default; SSRF-aware wrappers, egress allow-lists and IMDSv2 are spreading but
    //   opt-in) | legacy/10yr tech-debt HIGH (link-preview / image-proxy / import-URL
    //   features that pass raw user input straight to an HTTP client).
    // EXAMPLE: GET /ssrf/fetch?url=http://127.0.0.1:5001/access/admin/users
    //          -> SSRFs the app into requesting its own internal admin dump;
    //          real-world: url=http://169.254.169.254/latest/meta-data/iam/security-credentials/
    // FIX: allow-list scheme+host, resolve DNS and reject private/link-local IPs,
    //   disable redirects (see /ssrf/fetch-safe).
    // ========================================================================
    [HttpGet("/ssrf/fetch")]
    public async Task<IActionResult> Fetch(string url = "")
    {
        // VULNERABLE: HttpClient issues a GET to a fully user-controlled URL with
        // no scheme/host allow-list and no destination-IP validation.
        var (status, snippet, error) = await FetchSnippet(url);
        return new JsonResult(new
        {
            url,
            status,
            snippet,
            error,
            note = "the server, not the client, made this request — internal hosts and 169.254.169.254 are reachable"
        });
    }

    // ========================================================================
    // PERMUTATION 2 — blind SSRF via a user-supplied webhook / callback URL
    // OWASP 2021 A10 Server-Side Request Forgery -> folded into OWASP 2025 A01 Broken Access Control
    // CWE-918: Server-Side Request Forgery (SSRF)
    // EXPLOITATION LIKELIHOOD: HIGH — user-registered callback/webhook URLs are a
    //   first-class product feature, so the sink is exposed by design. Blind (the
    //   HTTP response is not shown to the caller) which lowers confirmability, but it
    //   still fires at internal services and http://169.254.169.254/latest/meta-data/;
    //   delivery-status/timing is a usable side channel and metadata exfil is severe.
    // PREVALENCE TODAY: greenfield MEDIUM (mature products validate webhook targets
    //   and pin egress, but many ship the feature with none of that) | legacy/10yr
    //   tech-debt HIGH (webhook/notification callers bolted on with no egress controls).
    // EXAMPLE: curl -X POST localhost:5001/ssrf/webhook -d 'url=http://169.254.169.254/latest/meta-data/'
    //          internal pivot: -d 'url=http://127.0.0.1:5001/access/admin/users'
    //          (JSON also accepted: -H 'Content-Type: application/json' -d '{"url":"..."}')
    // FIX: validate the callback destination (allow-list + resolved-IP block) before
    //   sending, and treat delivery status as sensitive (see /ssrf/fetch-safe).
    // ========================================================================
    [HttpPost("/ssrf/webhook")]
    public async Task<IActionResult> Webhook()
    {
        var url = await ReadUrlFromBody();
        if (string.IsNullOrWhiteSpace(url))
            return new JsonResult(new { error = "supply a webhook url via form field 'url'/'callback' or JSON {\"url\":...}" }) { StatusCode = 400 };

        // VULNERABLE: the caller-supplied callback URL is requested with no
        // validation. Blind SSRF — we deliberately do NOT return the response body,
        // only whether delivery connected, mirroring a real webhook sender.
        var (status, _, error) = await FetchSnippet(url);
        return new JsonResult(new
        {
            url,
            delivered = status > 0,
            status,
            error,
            note = "blind: the callback body is not returned; the request still reached the target (internal/metadata included)"
        });
    }

    // ========================================================================
    // PERMUTATION 3 — SSRF behind a NAIVE, bypassable substring blocklist
    // OWASP 2021 A10 Server-Side Request Forgery -> folded into OWASP 2025 A01 Broken Access Control
    // CWE-918: Server-Side Request Forgery (SSRF)
    // EXPLOITATION LIKELIHOOD: HIGH — the only defense is a substring check for
    //   'localhost'/'127.0.0.1', which is trivially bypassed. Deterministic and
    //   pre-auth; the block message even tells the attacker what to route around.
    // PREVALENCE TODAY: greenfield MEDIUM (teams that "add SSRF protection" reach for
    //   string/regex blocklists — a known-broken approach that still passes review) |
    //   legacy/10yr tech-debt HIGH (hand-rolled deny-lists on the raw URL string).
    // EXAMPLE: GET /ssrf/preview?url=http://127.0.0.1        -> blocked
    //          bypasses that all still hit loopback / link-local:
    //            http://127.0.0.1.nip.io   (DNS name resolving to 127.0.0.1)
    //            http://0177.0.0.1         (octal-encoded 127.0.0.1)
    //            http://2130706433         (decimal-encoded 127.0.0.1)
    //            http://[::1]              (IPv6 loopback — string never contains 127.0.0.1)
    // FIX: never blocklist the URL string; parse it, resolve DNS, and block the
    //   destination IP against private/link-local ranges (see /ssrf/fetch-safe).
    // ========================================================================
    [HttpGet("/ssrf/preview")]
    public async Task<IActionResult> Preview(string url = "")
    {
        // Naive, substring-only blocklist — the classic broken SSRF "defense".
        if (url.Contains("localhost") || url.Contains("127.0.0.1"))
            return new JsonResult(new { url, blocked = true, error = "blocked: local address not allowed" }) { StatusCode = 400 };

        // VULNERABLE: past the substring check, the raw URL is fetched — octal/decimal
        // IPs, DNS names like *.nip.io, and IPv6 [::1] all reach loopback/link-local.
        var (status, snippet, error) = await FetchSnippet(url);
        return new JsonResult(new
        {
            url,
            blocked = false,
            status,
            snippet,
            error,
            note = "substring blocklist bypassed via alternate host encodings (octal/decimal/nip.io/[::1])"
        });
    }

    // ========================================================================
    // SAFE REFERENCE — allow-list scheme+host AND block the resolved destination IP.
    // Shown so the vulnerable/safe pair can be diffed by SAST tooling and learners.
    //   (1) require an absolute http/https URL;
    //   (2) allow-list the host (reg-name), rejecting everything else;
    //   (3) resolve DNS and reject if ANY resolved address is loopback/private/
    //       link-local (127/8, 10/8, 172.16/12, 192.168/16, 169.254/16, ::1, fc00::/7)
    //       — this defeats decimal/octal/nip.io tricks and DNS-rebinding;
    //   (4) disable auto-redirects so a 30x cannot bounce us to an internal target.
    // ========================================================================
    [HttpGet("/ssrf/fetch-safe")]
    public async Task<IActionResult> FetchSafe(string url = "")
    {
        // Fixed allow-list of destinations this feature is actually permitted to reach.
        var allowedHosts = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "example.com", "www.example.com", "api.github.com"
        };

        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri))
            return new JsonResult(new { safe = true, url, error = "invalid absolute URL" }) { StatusCode = 400 };

        // SAFE (1)+(2): scheme + host allow-list.
        if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps)
            return new JsonResult(new { safe = true, url, error = "scheme not allowed (http/https only)" }) { StatusCode = 400 };
        if (!allowedHosts.Contains(uri.Host))
            return new JsonResult(new { safe = true, url, error = $"host '{uri.Host}' is not on the allow-list" }) { StatusCode = 400 };

        // SAFE (3): resolve and reject any private/link-local destination address.
        IPAddress[] addrs;
        try { addrs = await Dns.GetHostAddressesAsync(uri.Host); }
        catch (Exception e) { return new JsonResult(new { safe = true, url, error = "dns resolution failed: " + e.Message }) { StatusCode = 400 }; }

        foreach (var ip in addrs)
            if (IsPrivateOrLinkLocal(ip))
                return new JsonResult(new { safe = true, url, error = $"resolved to blocked address {ip}" }) { StatusCode = 400 };

        // SAFE (4): no auto-redirect (a 30x could otherwise point at an internal host).
        using var handler = new HttpClientHandler { AllowAutoRedirect = false };
        using var client = new HttpClient(handler) { Timeout = FetchTimeout };
        try
        {
            var resp = await client.GetAsync(uri);
            var body = await resp.Content.ReadAsStringAsync();
            return new JsonResult(new { safe = true, url, status = (int)resp.StatusCode, snippet = Snippet(body) });
        }
        catch (Exception e)
        {
            return new JsonResult(new { safe = true, url, error = e.Message }) { StatusCode = 502 };
        }
    }

    // --- helpers -------------------------------------------------------------

    // The dangerous sink shared by P1/P2/P3: fetch a user-controlled URL and return
    // (status, body-snippet, error). Auto-redirects follow by default (another SSRF
    // vector). Returns status 0 on transport/timeout failure.
    private static async Task<(int Status, string Snippet, string Error)> FetchSnippet(string url)
    {
        using var client = new HttpClient { Timeout = FetchTimeout };
        try
        {
            var resp = await client.GetAsync(url); // VULNERABLE: request issued to an unvalidated, user-controlled URL
            var body = await resp.Content.ReadAsStringAsync();
            return ((int)resp.StatusCode, Snippet(body), null);
        }
        catch (Exception e)
        {
            return (0, null, e.Message);
        }
    }

    private static string Snippet(string body)
    {
        if (string.IsNullOrEmpty(body)) return body;
        return body.Length > 512 ? body.Substring(0, 512) : body;
    }

    // Pull the target URL out of a form body ('url' or 'callback') or a JSON body.
    private async Task<string> ReadUrlFromBody()
    {
        if (Request.HasFormContentType)
        {
            var f = await Request.ReadFormAsync();
            var u = f["url"].ToString();
            if (!string.IsNullOrEmpty(u)) return u;
            var cb = f["callback"].ToString();
            if (!string.IsNullOrEmpty(cb)) return cb;
        }

        using var reader = new StreamReader(Request.Body);
        var raw = await reader.ReadToEndAsync();
        if (string.IsNullOrWhiteSpace(raw)) return "";
        try
        {
            var obj = JObject.Parse(raw);
            var u = (string)obj["url"] ?? (string)obj["callback"];
            if (!string.IsNullOrEmpty(u)) return u;
        }
        catch { /* not JSON — fall through and treat the raw body as the URL */ }
        return raw.Trim();
    }

    // True for loopback / private / link-local (incl. cloud-metadata 169.254.0.0/16)
    // addresses that a URL-fetching feature must never be allowed to reach.
    private static bool IsPrivateOrLinkLocal(IPAddress ip)
    {
        if (IPAddress.IsLoopback(ip)) return true; // 127.0.0.0/8, ::1

        if (ip.AddressFamily == AddressFamily.InterNetwork)
        {
            var b = ip.GetAddressBytes();
            if (b[0] == 0) return true;                                  // 0.0.0.0/8
            if (b[0] == 10) return true;                                 // 10.0.0.0/8
            if (b[0] == 172 && b[1] >= 16 && b[1] <= 31) return true;    // 172.16.0.0/12
            if (b[0] == 192 && b[1] == 168) return true;                 // 192.168.0.0/16
            if (b[0] == 169 && b[1] == 254) return true;                 // 169.254.0.0/16 (link-local + metadata)
            return false;
        }

        if (ip.AddressFamily == AddressFamily.InterNetworkV6)
        {
            if (ip.IsIPv6LinkLocal || ip.IsIPv6SiteLocal) return true;   // fe80::/10, fec0::/10
            var b = ip.GetAddressBytes();
            if ((b[0] & 0xFE) == 0xFC) return true;                      // fc00::/7 unique-local
            return false;
        }

        return false;
    }
}
